srv/modules/runners/upgrade.py

```diff
@@ -189,21 +217,71 @@ def status():
     __opts__['grains'] = __grains__
     __utils__ = salt.loader.utils(__opts__)
     __salt__ = salt.loader.minion_mods(__opts__, utils=__utils__)
+    master_minion = __salt__['master.minion']()
+
+    local = salt.client.LocalClient()
 
-    os_codename, _, ceph_version = __utils__['status.get_sys_versions']()
     # os_codename and ceph_version are pretty strings, e.g.:
     # - SUSE Linux Enterprise Server 15 SP1
     # - ceph version 14.2.1-468-g994fd9e0cc (994fd9e0cc[...]) nautilus (stable)
     # (Unless a node is down, in which case they'll be "Unknown (node down?)"
     # or "Not installed" if Ceph is not installed yet)
+    os_codename, _, ceph_version = __utils__['status.get_sys_versions']()
+
+    # Following is possibly a slightly obscure test.  We want to run a bunch of pre-upgrade
+    # checks when upgrading from SES6->7, but, this upgrade runner could also be invoked
+    # on a system which is part way through a SES5->6 upgrade, in which case some nodes
+    # would be SLE 12 SP3 and some would be SLE 15 SP1.  So, we check if there's any oses
+    # that aren't a variant on SLE 15 (or are down/Unkown).  If the list is empty, we
+    # assume we're on a SLE 15 SP1 or SP2 base, and can sensibly run the SES6->7 checks.
+    non_sle_15 = [os for os in set(os_codename.values())
+                    if not os.startswith("SUSE Linux Enterprise Server 15")
+                    and not os.startswith("Unknown")]
+    if len(non_sle_15) == 0:
+        require_osd_release = list(local.cmd(master_minion, 'cmd.run',
+            ["ceph osd dump|awk '/require_osd_release/ {print $2}'"]).items())[0][1]
```

Comment:
Yikes.

I searched for a better way to retrieve `require_osd_release` and did not find anything; what a shame.

What about moving that to a separate module? Something like.

`ceph.require_osd_release` which could be executed on the master or any mon/admin node


Alternatively we may use `--json-pretty` and pipe it in `jq`. Sounds a bit less error prone to me?

Thoughts?

Reply:
Yeah, I've shifted this to the osd module (meaning we can talk to rados directly and not screw around with `awk` or `jq`). This also means it's usable from the adopt state, so we can abort any adoption if this setting isn't correct.